File: sdk/js/src/diagnostic.ts
/**
 * One thing the engine noticed about a document.
 *
 * Passed through, never interpreted. `code` and `args` are the engine's frozen
 * contract — a translating consumer renders its own message from them — so this
 * class parses the wire and stops. It does not translate, it does not
 * re-classify, and it never becomes an exception: a render that warns still
 * succeeded, and a render that failed says why in these.
 */

/** One engine diagnostic, exactly as the engine stated it. */
export class Diagnostic {
  readonly severity: string | null;
  readonly code: string | null;
  readonly category: string | null;
  readonly message: string | null;
  readonly path: string | null;
  readonly args: Record<string, unknown>;
  readonly origin: string | null;

  constructor(item: Record<string, unknown>) {
    this.severity = text(item.severity);
    this.code = text(item.code);
    this.category = text(item.category);
    this.message = text(item.message);
    this.path = text(item.path);
    this.args = record(item.args);
    this.origin = text(item.origin);
  }

  /** Every diagnostic in a payload, or nothing at all for an empty one. */
  static parse(payload: string): Diagnostic[] {
    if (!payload) {
      return [];
    }

    const items = record(JSON.parse(payload)).items;
    return Array.isArray(items) ? items.map((item) => new Diagnostic(record(item))) : [];
  }

  /** A NOUN takes `is`, so this does not read as "the error object". */
  get isError(): boolean {
    return this.severity === 'error';
  }

  get isWarning(): boolean {
    return this.severity === 'warning';
  }

  toString(): string {
    return [this.path, this.message].filter((part) => part !== null).join(': ');
  }
}

function text(value: unknown): string | null {
  return typeof value === 'string' ? value : null;
}

/**
 * A parsed payload as a plain map.
 *
 * `Object.create(null)`-free on purpose: what matters is that a `__proto__` key
 * arriving from `JSON.parse` is an ORDINARY own property (the parser never runs
 * the setter), and that every lookup this package makes against attacker text
 * goes through a `Map` rather than through an object it could inherit from.
 */
function record(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
}

File: sdk/python/src/shojiku/diagnostic.py
"""One thing the engine noticed about a document.

Passed through, never interpreted. ``code`` and ``args`` are the engine's
frozen contract — a translating consumer renders its own message from them — so
this class parses the wire and stops. It does not translate, it does not
re-classify, and it never becomes an exception: a render that warns still
succeeded, and a render that failed says why in these.
"""

from __future__ import annotations

import json
from typing import Any


class Diagnostic:
    """One engine diagnostic, exactly as the engine stated it."""

    def __init__(self, item: dict[str, Any]) -> None:
        self.severity: str | None = item.get("severity")
        self.code: str | None = item.get("code")
        self.category: str | None = item.get("category")
        self.message: str | None = item.get("message")
        self.path: str | None = item.get("path")
        self.args: dict[str, Any] = item.get("args") or {}
        self.origin: str | None = item.get("origin")

    @staticmethod
    def parse(payload: str) -> list[Diagnostic]:
        """Every diagnostic in a payload, or nothing at all for an empty one."""
        if not payload:
            return []

        items = json.loads(payload).get("items")
        return [Diagnostic(item) for item in items or []]

    @property
    def is_error(self) -> bool:
        return self.severity == "error"

    @property
    def is_warning(self) -> bool:
        return self.severity == "warning"

    def __str__(self) -> str:
        return ": ".join(part for part in (self.path, self.message) if part is not None)
